File: tui/reducer.go
package tui

type Graph struct {
	Nodes       map[string]Node
	Edges       map[string]Edge
	established map[string]bool
	statusRev   map[string]uint64
	tombstones  map[string]uint64
}

func EmptyGraph() Graph {
	return Graph{
		Nodes:       make(map[string]Node),
		Edges:       make(map[string]Edge),
		established: make(map[string]bool),
		statusRev:   make(map[string]uint64),
		tombstones:  make(map[string]uint64),
	}
}

func applyStatusFields(target, source *Node) {
	if source.Status != "" {
		target.Status = source.Status
	}
	if source.TEnd != nil {
		target.TEnd = source.TEnd
	}
	if source.DurationMS != nil {
		target.DurationMS = source.DurationMS
	}
}

func Apply(g *Graph, ev SseEvent) {
	switch ev.Kind {
	case "node_upsert":
		applyNodeUpsert(g, ev)
	case "node_status":
		applyNodeStatus(g, ev)
	case "node_merge":
		applyNodeMerge(g, ev)
	case "edge_upsert":
		applyEdgeUpsert(g, ev)
	case "edge_delete":
		applyEdgeDelete(g, ev)
	}
}

func applyNodeUpsert(g *Graph, ev SseEvent) {
	if ev.Node == nil {
		return
	}
	n := ev.Node
	tombKey := "node:" + n.ID
	if tomb, ok := g.tombstones[tombKey]; ok && n.Rev <= tomb {
		return
	}
	existing, exists := g.Nodes[n.ID]
	if exists && g.established[n.ID] {
		if n.Rev <= existing.Rev {
			return
		}
	}
	if exists && !g.established[n.ID] {
		merged := *n
		prevStatusRev := g.statusRev[n.ID]
		if prevStatusRev > n.Rev {
			if existing.Status != "" {
				merged.Status = existing.Status
			}
			if existing.TEnd != nil {
				merged.TEnd = existing.TEnd
			}
			if existing.DurationMS != nil {
				merged.DurationMS = existing.DurationMS
			}
		}
		g.Nodes[n.ID] = merged
	} else {
		g.Nodes[n.ID] = *n
	}
	g.established[n.ID] = true
	prev := g.statusRev[n.ID]
	if n.Rev >= prev {
		g.statusRev[n.ID] = n.Rev
	}
}

func applyNodeStatus(g *Graph, ev SseEvent) {
	if ev.Node == nil {
		return
	}
	patch := ev.Node
	tombKey := "node:" + patch.ID
	if tomb, ok := g.tombstones[tombKey]; ok && patch.Rev <= tomb {
		return
	}
	prevStatusRev := g.statusRev[patch.ID]
	existing, exists := g.Nodes[patch.ID]
	if !exists {
		seed := Node{
			ID:    patch.ID,
			RunID: patch.RunID,
			Type:  patch.Type,
			Rev:   patch.Rev,
		}
		applyStatusFields(&seed, patch)
		g.Nodes[patch.ID] = seed
		g.statusRev[patch.ID] = patch.Rev
		return
	}
	if patch.Rev >= prevStatusRev {
		applyStatusFields(&existing, patch)
		g.statusRev[patch.ID] = patch.Rev
		if !g.established[patch.ID] {
			existing.Rev = patch.Rev
		}
		g.Nodes[patch.ID] = existing
	}
}

func applyNodeMerge(g *Graph, ev SseEvent) {
	if ev.Node == nil {
		return
	}
	n := ev.Node
	if ev.OldID != "" && ev.OldID != n.ID {
		delete(g.Nodes, ev.OldID)
		delete(g.established, ev.OldID)
		delete(g.statusRev, ev.OldID)
		tombKey := "node:" + ev.OldID
		existing := g.tombstones[tombKey]
		if ev.Rev > existing {
			g.tombstones[tombKey] = ev.Rev
		}
		for id, e := range g.Edges {
			changed := false
			if e.Src == ev.OldID {
				e.Src = n.ID
				changed = true
			}
			if e.Dst == ev.OldID {
				e.Dst = n.ID
				changed = true
			}
			if changed {
				g.Edges[id] = e
			}
		}
	}
	g.Nodes[n.ID] = *n
	g.established[n.ID] = true
	g.statusRev[n.ID] = n.Rev
}

func applyEdgeUpsert(g *Graph, ev SseEvent) {
	if ev.Edge == nil {
		return
	}
	e := ev.Edge
	if tomb, ok := g.tombstones[e.ID]; ok && e.Rev <= tomb {
		return
	}
	if existing, ok := g.Edges[e.ID]; ok && e.Rev <= existing.Rev {
		return
	}
	g.Edges[e.ID] = *e
}

func applyEdgeDelete(g *Graph, ev SseEvent) {
	if ev.Edge == nil {
		return
	}
	id := ev.Edge.ID
	existing := g.tombstones[id]
	if ev.Rev > existing {
		g.tombstones[id] = ev.Rev
	}
	delete(g.Edges, id)
}

File: webui/web/src/lib/reducer/reducer.ts
import type { Node, Edge, SseEvent } from '../types';

export interface GraphState {
  nodes: Record<string, Node>;
  edges: Record<string, Edge>;
  established: Record<string, true>;
  tombstones: Record<string, number>;
  // Internal: tracks the highest rev of any event that carried status fields
  // for each node. Kept separate from nodes[id].rev so that status LWW is
  // independent of the upsert identity guard.
  statusRevs: Record<string, number>;
}

export function emptyState(): GraphState {
  return { nodes: {}, edges: {}, established: {}, tombstones: {}, statusRevs: {} };
}

function applyStatusFields(target: Node, source: Partial<Node>): void {
  if (source.status !== undefined) target.status = source.status;
  if (source.t_end !== undefined) target.t_end = source.t_end;
  if (source.duration_ms !== undefined) target.duration_ms = source.duration_ms;
}

export function applyDelta(state: GraphState, ev: SseEvent): void {
  switch (ev.kind) {
    case 'node_upsert': {
      if (!ev.node) return;
      const n = ev.node;
      const nodeTomb = state.tombstones[`node:${n.id}`];
      if (nodeTomb !== undefined && n.rev <= nodeTomb) return;
      const existing = state.nodes[n.id];
      if (existing && state.established[n.id]) {
        if (n.rev <= existing.rev) return;
      }
      if (existing && !state.established[n.id]) {
        // Merging a new upsert into an unestablished (status-seeded) node.
        // Identity fields come from the upsert; status fields are latest-rev-wins
        // between the upsert rev and the previously tracked statusRevs[id].
        // statusRevs[n.id] is always defined here: an unestablished node can only
        // exist because node_status seeded it, and node_status always sets statusRevs.
        const merged: Node = { ...n };
        const prevStatusRev = state.statusRevs[n.id] as number;
        if (prevStatusRev > n.rev) {
          // Keep the higher-rev status fields from the seed.
          if (existing.status !== undefined) merged.status = existing.status;
          if (existing.t_end !== undefined) merged.t_end = existing.t_end;
          if (existing.duration_ms !== undefined) merged.duration_ms = existing.duration_ms;
        }
        state.nodes[n.id] = merged;
      } else {
        state.nodes[n.id] = { ...n };
      }
      state.established[n.id] = true;
      // An upsert is also a status-carrying event at its rev.
      const prevUpsertStatusRev = state.statusRevs[n.id] ?? 0;
      if (n.rev >= prevUpsertStatusRev) {
        state.statusRevs[n.id] = n.rev;
      }
      return;
    }

    case 'node_status': {
      if (!ev.node) return;
      const patch = ev.node;
      const nodeTomb = state.tombstones[`node:${patch.id}`];
      if (nodeTomb !== undefined && patch.rev <= nodeTomb) return;
      const existing = state.nodes[patch.id];
      // Latest-rev-wins for status fields, independently of the upsert rev.
      const prevStatusRev = state.statusRevs[patch.id] ?? 0;
      if (!existing) {
        const seed: Node = {
          id: patch.id,
          run_id: patch.run_id,
          type: patch.type,
          rev: patch.rev,
        };
        applyStatusFields(seed, patch);
        state.nodes[patch.id] = seed;
        state.statusRevs[patch.id] = patch.rev;
      } else {
        if (patch.rev >= prevStatusRev) {
          applyStatusFields(existing, patch);
          state.statusRevs[patch.id] = patch.rev;
          if (!state.established[patch.id]) {
            existing.rev = patch.rev;
          }
        }
      }
      return;
    }

    case 'node_merge': {
      if (!ev.node) return;
      const n = ev.node;
      if (ev.old_id && ev.old_id !== n.id) {
        delete state.nodes[ev.old_id];
        delete state.established[ev.old_id];
        delete state.statusRevs[ev.old_id];
        const existingTomb = state.tombstones[`node:${ev.old_id}`];
        state.tombstones[`node:${ev.old_id}`] =
          existingTomb !== undefined ? Math.max(existingTomb, ev.rev) : ev.rev;
        for (const e of Object.values(state.edges)) {
          if (e.src === ev.old_id) e.src = n.id;
          if (e.dst === ev.old_id) e.dst = n.id;
        }
      }
      state.nodes[n.id] = { ...n };
      state.established[n.id] = true;
      state.statusRevs[n.id] = n.rev;
      return;
    }

    case 'edge_upsert': {
      if (!ev.edge) return;
      const e = ev.edge;
      // Backend assigns monotonically increasing revs, so delete.rev > upsert.rev
      // always holds for any given edge in production. The tombstone LWW is correct
      // for all backend-realistic inputs: if a delete arrived first (out of SSE order),
      // its higher rev is already in the tombstone and will block this upsert.
      const tomb = state.tombstones[e.id];
      if (tomb !== undefined && e.rev <= tomb) return;
      const existing = state.edges[e.id];
      if (existing && e.rev <= existing.rev) return;
      state.edges[e.id] = { ...e };
      return;
    }

    case 'edge_delete': {
      if (!ev.edge) return;
      const id = ev.edge.id;
      // Backend assigns monotonically increasing revs, so a delete always has a
      // strictly higher rev than the same edge's prior upsert. Both arrival orders
      // therefore converge: delete-first records the tombstone which then blocks the
      // later upsert (upsert.rev < tombstone); upsert-first writes the edge which is
      // then removed here. The case upsert.rev > delete.rev cannot arise from the
      // backend and is out of scope.
      const existing = state.tombstones[id];
      state.tombstones[id] = existing !== undefined ? Math.max(existing, ev.rev) : ev.rev;
      delete state.edges[id];
      return;
    }

    default:
      return;
  }
}
